### PR TITLE
Experiment/interaction state layer

### LINKED ISSUE #2008: Refactor interaction state solution and apply hover
📣 Update: Implement state layer based solution

~~Other than `<button>` which components and elements should utilize the `apply-hover()` mixin?~~

~~A quick search for `:hover` says it's only the [calendar component](https://github.com/kirbydesign/designsystem/blob/45f0fd17d7f996319ab8ff5dff8c6e24f12f10df/libs/designsystem/src/lib/components/calendar/calendar.component.scss) and the [chip component](https://github.com/kirbydesign/designsystem/blob/45f0fd17d7f996319ab8ff5dff8c6e24f12f10df/libs/designsystem/src/lib/components/chip/chip.component.scss).~~

The chip components is (apparently) the only place we use the `theme-background()` mixin. Consider if there's room for improvement, i.e., can it be replaced by using the `apply-hover()` mixin?

~~Consider splitting this issue into at least 3 separate issues: Calender, Chip and Item + Dropdown.~~

### Button 
- [x] Apply state layer based solution to the kirby button directive.

### Calendar

- [x] Apply hover to `.day` elements
  - Must handle `selected` and `today` also
- [x] Check year navigator button (fixed by Dropdown?)
- [x] Check month navigator buttons
- Hover color on all of the above should be the same (consult UX)

### Chip

- [x] Remove use of `theme-background()` 
- [x] Apply hover using the `apply-hover()` mixin

### Item and Dropdown

- [x] Refactor hover on Items in Card in Dropdown
  - The hover effect is currently controlled by `ion-item`
  - Consider if `apply-hover()` can/should be used
  - ⚠️ What are the consequences if we can't use `apply-hover()`?
    - Do we want to use 3 (or more) different solutions for interaction states? 
    - Solution: Use separate mixin for Ionic components (part of current refactorization)
- Clean up old API/interface
  - [x] Remove `--kirby-item-background-hover` state related CSS custom properties
  - [x] Update Cookbook showcase

### Other Components

- [x] Page - Toolbar buttons etc.
- [x] Segmented Control - fix what doesn't get fixed by applying state layer to Chip
- [x] Slide Button
- [x] #2056
- [x] #2057
- [x] Improve suboptimal solutions for `ion-back-button` and `ion-segment-button` (in Page and Segmented Control)
- [x] FAB Button: Ionic variant of solution + transition mixin
- [x] #2016 

### Refactor interaction state solution and apply hover - Part 2

See #2028

- #2058
- #2089
- #2090
- #2091
- #2092

### Misc.

- [x] ♿️ Accessibility: Hide state layer
- [x] 🔲 Borders: Should state layer cover border or not?
  - See e.g. button attention level 3 vs other attention levels
- [x] Verify that there are no other places we should replace `:hover` with `apply-hover()`
  - ion-reorder - maybe later
- [x] 🚶 Walk through all hover cases with UX/UI - adjust on the fly as needed 
- [ ] (Optional) Move the `hover()` mixin to `libs/core/src/scss/interaction-state/_hover.scss`

**Note:** If this issue is a can of worms in terms of rewriting chip scss & theme-background mixin - it should be split into a seperate issue.

### PR DESCRIPTION
**Pull request by jkaltoft**
# ⚠️ This is an experiment. Do not merge PR ⚠️

## Which issue does this PR close?

This PR closes #2008 

...but it is also a refactorization of interaction state layer solution in general

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [ ] Request that the changes are code-reviewed 
- [ ] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be automatically merged to master via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

### CODE REVIEW

File: libs/designsystem/src/lib/components/button/button.component.scss [hunk 1]
```diff
@@ -119,18 +87,28 @@ $button-width: (
...
 
+  @include interaction-state.setup-layer;
+  @include interaction-state.apply-hover;
+  @include interaction-state.extend-content {
+    display: inherit;
+    flex-direction: inherit;
+    align-items: inherit;
+    justify-content: inherit;
+    padding-inline: var(--kirby-button-padding-left) var(--kirby-button-padding-right);
+  }
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/button/button.component.scss:108]: What's the point of `extend-content`? Why not just write: 
```sass
.content {
    display: inherit;
    flex-direction: inherit;
    align-items: inherit;
    justify-content: inherit;
    padding-inline: var(--kirby-button-padding-left) var(--kirby-button-padding-right);
}
```

**Review by MadsBuchmann — COMMENTED** [review]
I've been through the code - i left a few questions :D

File: libs/designsystem/src/lib/components/button/button.component.integration.spec.ts [hunk 2]
```diff
@@ -414,9 +414,10 @@ describe('ButtonComponent configured with icon only', () => {
 });
 
 describe('ButtonComponent configured with text and icon', () => {
-  let kirbyIcon: Element;
   let spectator: SpectatorHost<ButtonComponent>;
   let element: HTMLButtonElement;
+  let kirbyIcon: Element;
```
[the remark this answers is not in the document]
Comment (jkaltoft) [reply libs/designsystem/src/lib/components/button/button.component.integration.spec.ts:419]: I don't know. All I did was reorder the lines 🙂

File: libs/core/src/scss/interaction-state/_layer.scss [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      opacity: var(--state-layer-opacity);
+      background-color: var(--state-layer-background-color);
+
+      // TODO: Extract transition into "interaction state trasition" mixin
+      // transition-property: all;
+      // transition-duration: 180ms;
+      // transition-delay: 5ms;
+      transition-property: all;
+      transition-duration: utils.get-transition-duration('quick');
+      transition-timing-function: utils.get-transition-easing('static');
```
[the remark this answers is not in the document]
Comment (jkaltoft) [reply libs/core/src/scss/interaction-state/_layer.scss:50]: Yes.

There's already a TODO for abstracting away the transition declaration:

> `// TODO: Extract transition into "interaction state trasition" mixin`

It should probably not be configurable, though

File: libs/designsystem/src/lib/components/button/button.component.scss (hunk 1 above)
Comment (jkaltoft) [reply libs/designsystem/src/lib/components/button/button.component.scss:108]: It should not be applied to `.content` for all interaction state layers. And the `.content` rule block should not be applied directly in a component - hence the abstraction.

**Pull request closed without merging**